libc/string/string_tokpbrkstr.c

```diff
@@ -0,0 +1,875 @@
+#define EXTENDED_ASCII_LENGTH 256
+#define BUFFSIZE              24
+#define LOREM_IPSUM           "Lorem Ipsum Dolor"
+
+TEST_GROUP(string_tok);
+TEST_GROUP(string_tok_r);
+TEST_GROUP(string_str);
+TEST_GROUP(string_pbrk);
+
+static char *create_extAscii_set(void)
+{
+	int i;
+	char *hold = malloc(EXTENDED_ASCII_LENGTH);
+
+	for (i = 1; i < EXTENDED_ASCII_LENGTH; i++) {
+		hold[i - 1] = i;
+	}
+	hold[i - 1] = 0;
+
+	return hold;
+}
+
+
+TEST_SETUP(string_tok)
+{
+}
+
+
+TEST_TEAR_DOWN(string_tok)
+{
+}
+
+
+TEST(string_tok, basic)
+{
+	char str[] = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas eleifend elementum tellum.";
+	const char *separators = ", .";
+	char *token;
+
+	TEST_ASSERT_EQUAL_PTR(str, (token = strtok(str, separators)));
+	TEST_ASSERT_EQUAL_STRING("Lorem", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[6], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("ipsum", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[12], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("dolor", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[18], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("sit", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[22], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("amet", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[28], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("consectetur", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[40], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("adipiscing", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[51], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("elit", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[57], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("Maecenas", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[66], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("eleifend", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[75], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("elementum", token);
+
+	TEST_ASSERT_EQUAL_PTR(&str[85], (token = strtok(NULL, separators)));
+	TEST_ASSERT_EQUAL_STRING("tellum", token);
+
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, separators));
+}
+
+
+/* assert that looking for tokens stops after encountering even one NUL term */
+TEST(string_tok, torn)
+{
+	char str[] = "Lor\0em ipsum";
+	const char *separators = " ";
+	char *tok;
+	TEST_ASSERT_EQUAL_PTR(str, (tok = strtok(str, separators)));
+	TEST_ASSERT_EQUAL_STRING("Lor", tok);
+	TEST_ASSERT_EQUAL_PTR(NULL, (tok = strtok(NULL, separators)));
+}
+
+
+TEST(string_tok, empty_args)
+{
+	char empty[] = { 0 },
+		 str[] = "abc";
+
+	/* Different scenarios of acquiring null pointer as output*/
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok("", "d"));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, ""));
+
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(empty, ""));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, ""));
+
+	TEST_ASSERT_EQUAL_PTR(str, strtok(str, empty));
+	TEST_ASSERT_EQUAL_PTR(NULL, strtok(NULL, empty));
```

Comment:
I applied those changes but in suggestions is an error because, with the string passed by hand, we can't assert PTR and one line above instead of passing string as `""` i was forced to use `"\0"` because of restrictions in variables passed to strtok(GCC)

![image](https://github.com/phoenix-rtos/phoenix-rtos-tests/assets/49757239/9ca8ff46-c1e1-4895-9d3e-6a129f6e9b0e)

Reply:
oh, ok there is `restrict` in the declaration. Using "\0" seems to be fine in this case.